validate: mismatch between NXdata@axes and available fields

BTW, I'm surprised that `punx validate` did not identify the very real problem with this NeXus file:

```
    data:NXdata
      @NX_class = "NXdata"
      @axes = ["und"]
      @signal = "XPS_sample"
      @target = "/entry/data"
      EPOCH --> /entry/instrument/bluesky/streams/primary/und_readback/time
      M6_collector --> /entry/instrument/bluesky/streams/primary/M6_collector/value
      XPS_pd --> /entry/instrument/bluesky/streams/primary/XPS_pd/value
      XPS_sample --> /entry/instrument/bluesky/streams/primary/XPS_sample/value
      und_readback --> /entry/instrument/bluesky/streams/primary/und_readback/value
```

In NXdata, the `axes` attribute value names a field that exists in the group (either as HDF5 dataset or HDF5 link).  And that is clearly not true here.

_Originally posted by @prjemian in https://github.com/BCDA-APS/apstools/issues/806#issuecomment-1433619315_

This test method:

```py
def test_i219(tempdir):
    h5file = tempdir / "test_file.h5"
    assert not h5file.exists()

    with h5py.File(h5file, "w") as root:
        root.attrs["default"] = "entry"

        nxentry = root.create_group(root.attrs["default"])
        nxentry.attrs["NX_class"] = "NXentry"
        nxentry.attrs["default"] = "data"

        nxdata = nxentry.create_group(nxentry.attrs["default"])
        nxdata.attrs["NX_class"] = "NXdata"

        # these match
        nxdata.attrs["signal"] = "XPS_sample"
        nxdata.create_dataset("XPS_sample", data=[1, 2, 3])
        assert nxdata.attrs["signal"] in nxdata

        # these do not match
        nxdata.attrs["axes"] = ["und"]
        nxdata.create_dataset("und_readback", data=[3, 4, 1])
        for k in nxdata.attrs["axes"]:
            assert k not in nxdata

    assert h5file.exists()

    validator = validate.Data_File_Validator()
    assert isinstance(validator, validate.Data_File_Validator)

    validator.validate(h5file)

    average = validator.finding_score()[-1]
    assert average < -10_000
```

confirms that validate fails to catch this mismatch:

```py
        average = validator.finding_score()[-1]
>       assert average < -10_000
E       assert 98.84415584415585 < -10000

punx/tests/test_i219_nxdata_mismatch.py:60: AssertionError
=========================== short test summary info ============================
FAILED punx/tests/test_i219_nxdata_mismatch.py::test_i219 - assert 98.8441558...
============================== 1 failed in 1.11s ===============================
```

With any single test failure, `average` should be a negative number.

--- a/punx/finding.py
+++ b/punx/finding.py
@@ -25,6 +25,7 @@ import hashlib
 VALID_STATUS_DICT = {}
 """dictionary (by names) of all available validations"""
 
+
 class ValidationResultStatus(object):
     """
     summary result of a Finding

--- a/punx/finding.py
+++ b/punx/finding.py
@@ -22,6 +22,9 @@ document each item during validation
 import hashlib
 
 
+VALID_STATUS_DICT = {}
+"""dictionary (by names) of all available validations"""
+
 class ValidationResultStatus(object):
     """
     summary result of a Finding
@@ -36,6 +39,7 @@ class ValidationResultStatus(object):
         self.value = value
         self.color = color
         self.description = description
+        VALID_STATUS_DICT[key] = self
 
     def __str__(self, *args, **kwargs):
         return self.key
@@ -61,9 +65,7 @@ OPTIONAL = ValidationResultStatus(
     "OPTIONAL", 99, "grey", "allowed by NeXus specification, not identified"
 )
 
-VALID_STATUS_LIST = (OK, NOTE, WARN, ERROR, TODO, UNUSED, COMMENT, OPTIONAL)
-VALID_STATUS_DICT = {str(f): f for f in VALID_STATUS_LIST}
-"""dictionary (by names) of all available validations"""
+VALID_STATUS_LIST = tuple(VALID_STATUS_DICT.values())
 
 TF_RESULT = {True: OK, False: ERROR}
 

--- a/punx/tests/test_i219_nxdata_mismatch.py
+++ b/punx/tests/test_i219_nxdata_mismatch.py
@@ -1,0 +1,60 @@
+"""
+Validate mismatch between NXdata@axes and available fields.
+"""
+
+import pathlib
+import shutil
+import tempfile
+
+import h5py
+import pytest
+
+from .. import validate
+
+
+@pytest.fixture(scope="function")
+def tempdir():
+    """EPICS Motor"""
+    path = pathlib.Path(tempfile.mkdtemp())
+    assert path.exists()
+
+    yield path
+
+    if path.exists():
+        shutil.rmtree(path)
+
+
+def test_i219(tempdir):
+    h5file = tempdir / "test_file.h5"
+    assert not h5file.exists()
+
+    with h5py.File(h5file, "w") as root:
+        root.attrs["default"] = "entry"
+
+        nxentry = root.create_group(root.attrs["default"])
+        nxentry.attrs["NX_class"] = "NXentry"
+        nxentry.attrs["default"] = "data"
+
+        nxdata = nxentry.create_group(nxentry.attrs["default"])
+        nxdata.attrs["NX_class"] = "NXdata"
+
+        # these match
+        nxdata.attrs["signal"] = "XPS_sample"
+        nxdata.create_dataset("XPS_sample", data=[1, 2, 3])
+        assert nxdata.attrs["signal"] in nxdata
+
+        # these do not match
+        nxdata.attrs["axes"] = ["und"]
+        nxdata.create_dataset("und_readback", data=[3, 4, 1])
+        for k in nxdata.attrs["axes"]:
+            assert k not in nxdata
+
+    assert h5file.exists()
+
+    validator = validate.Data_File_Validator()
+    assert isinstance(validator, validate.Data_File_Validator)
+
+    validator.validate(h5file)
+
+    average = validator.finding_score()[-1]
+    assert average < -10_000

--- a/punx/tests/test_i219_nxdata_mismatch.py
+++ b/punx/tests/test_i219_nxdata_mismatch.py
@@ -14,7 +14,7 @@ from .. import validate
 
 @pytest.fixture(scope="function")
 def tempdir():
-    """EPICS Motor"""
+    """Temporary directory for testing."""
     path = pathlib.Path(tempfile.mkdtemp())
     assert path.exists()
 

--- a/punx/tests/test_validate.py
+++ b/punx/tests/test_validate.py
@@ -18,19 +18,13 @@ ISSUES
 * [ ] #72  refactor to validate application definitions
 """
 
-import h5py
 import os
+
+import h5py
 import pytest
 
-from ._core import DEFAULT_NXDL_FILE_SET
-from ._core import EXAMPLE_DATA_DIR
-from ._core import hfile
-from ._core import No_Exception
-from .. import FileNotFound
-from .. import finding
-from .. import HDF5_Open_Error
-from .. import utils
-from .. import validate
+from .. import FileNotFound, HDF5_Open_Error, finding, utils, validate
+from ._core import DEFAULT_NXDL_FILE_SET, EXAMPLE_DATA_DIR, No_Exception, hfile
 
 
 def avert_exception(fname):
@@ -103,7 +97,7 @@ def test_valid_nexus_file_constructed(hfile):
         [True, "a4fd52d", 1],
         [True, "v3.3", 0],
         [False, "no_such_file_set", None],
-    ]
+    ],
 )
 def test_NXdata_requirement_or_optional(valid_ref, file_set_name, count, hfile):
     """
@@ -553,7 +547,7 @@ def test_default_plot_v3_pass_multi(hfile):
         [DEFAULT_NXDL_FILE_SET, finding.NOTE, 1],
         ["a4fd52d", finding.NOTE, 0],
         ["v3.3", finding.ERROR, 0],
-    ]
+    ],
 )
 def test_default_plot_v3_fail(file_set_name, status, occurs, hfile):
     setup_simple_test_file_default_plot(hfile)
@@ -605,7 +599,7 @@ def test_default_plot_v2_pass(hfile):
         [DEFAULT_NXDL_FILE_SET, finding.NOTE, 1],
         ["a4fd52d", finding.NOTE, 0],
         ["v3.3", finding.ERROR, 0],
-    ]
+    ],
 )
 def test_default_plot_v2_fail_no_signal(file_set_name, status, occurs, hfile):
     setup_simple_test_file_default_plot(hfile)
@@ -632,7 +626,7 @@ def test_default_plot_v2_fail_no_signal(file_set_name, status, occurs, hfile):
         [DEFAULT_NXDL_FILE_SET, finding.NOTE, 1],
         ["a4fd52d", finding.NOTE, 0],
         ["v3.3", finding.ERROR, 0],
-    ]
+    ],
 )
 def test_default_plot_v2_fail_multi_signal(file_set_name, status, occurs, hfile):
     setup_simple_test_file_default_plot(hfile)

--- a/punx/tests/test_validate.py
+++ b/punx/tests/test_validate.py
@@ -681,8 +681,11 @@ def test_default_plot_v2_fail_multi_signal(file_set_name, status, occurs, hfile)
         ["writer_2_1.hdf5", "note", 0],
         ["writer_2_1.hdf5", "TODO", 11],
         ["1998spheres.h5", "ERROR", 2],
+        ["02_03_setup.h5", "NOTE", 98],
+        ["02_03_setup.h5", "OPTIONAL", 70],
+        ["02_03_setup.h5", "ERROR", 0],
         ["02_03_setup.h5", "NOTE,OPTIONAL,ERROR", 98 + 70 + 0],
-        ["prj_test.nexus.hdf5", "", 121],
+        ["prj_test.nexus.hdf5", "", 123],
     ],
 )
 def test_report_option(infile, report, observations, capsys):

--- a/punx/validations/attribute.py
+++ b/punx/validations/attribute.py
@@ -8,10 +8,10 @@
 # The full license is in the file LICENSE.txt, distributed with this software.
 # -----------------------------------------------------------------------------
 
-from .. import finding
-from .. import utils
-from . import item_name
+import numpy
 
+from .. import finding, utils
+from . import item_name
 
 TEST_NAME = "attribute value"
 
@@ -25,18 +25,7 @@ def verify(validator, v_item):
     if v_item.h5_address.find("@") < 0:
         return
 
-    special_handlers = {
-        "NX_class": nxclass_handler,
-        "target": target_handler,
-        "default": generic_handler,
-        "signal": signal_handler,
-        "axes": axes_handler,
-        "axis": generic_handler,
-        "primary": generic_handler,
-        "units": units_handler,
-    }
-
-    handler = special_handlers.get(v_item.name) or generic_handler
+    handler = HANDLER_DICT.get(v_item.name) or generic_handler
     handler(validator, v_item)
 
 
@@ -70,14 +59,37 @@ def axes_handler(validator, v_item):
     """
     validate @axes
     """
-    # TODO: axes_attr = v_item.h5_object
-    # if this is not an array, make it axes_attr_array
+    axes_attr = v_item.h5_object
+    if not isinstance(axes_attr, (list, numpy.ndarray)):
+        axes_attr = [axes_attr]  # ensure it is array
 
-    # TODO: need to know shape of signal data
-    # TODO: compare len(axes_attr_array) with range of signal data
-    # TODO: check each value of array that is a validItemName and points to actual local field
+    for axis_name in utils.decode_byte_string(v_item.h5_object):
+        # check each value of array that is a validItemName
+        k = item_name.validItemName_match_key(validator, axis_name)
+        test_name = f"valid name @axes['{axis_name}']"
+        if k is None:
+            status = finding.ERROR
+            k = "not a valid NeXus name"
+        else:
+            if k.startswith("strict"):
+                status = finding.OK
+            else:
+                status = finding.NOTE
+        validator.record_finding(v_item, test_name, status, k)
 
-    generic_handler(validator, v_item)
+        # check axis_name names local field (in this group)
+        test_name = f"axes['{axis_name}'] exists"
+        h5parent = v_item.parent.h5_object
+        if h5parent.get(axis_name) is not None:
+            status = finding.OK
+            k = "found field for named axis"
+        else:
+            status = finding.ERROR
+            k = "did not find field for named axis"
+        validator.record_finding(v_item, test_name, status, k)
+
+    test_name = "compare shapes of signal & axes data"
+    validator.record_finding(v_item, TEST_NAME, finding.TODO, "implement")
 
 
 def nxclass_handler(validator, v_item):
@@ -191,3 +203,16 @@ def generic_handler(validator, v_item):
     if v_item.name.endswith("_indices"):
         pass
     validator.record_finding(v_item, TEST_NAME, finding.TODO, "implement")
+
+
+# must define after the handler functions
+HANDLER_DICT = {
+    "NX_class": nxclass_handler,
+    "target": target_handler,
+    "default": generic_handler,
+    "signal": signal_handler,
+    "axes": axes_handler,
+    "axis": generic_handler,
+    "primary": generic_handler,
+    "units": units_handler,
+}

--- a/punx/validations/attribute.py
+++ b/punx/validations/attribute.py
@@ -59,11 +59,11 @@ def axes_handler(validator, v_item):
     """
     validate @axes
     """
-    axes_attr = v_item.h5_object
+    axes_attr = utils.decode_byte_string(v_item.h5_object)
     if not isinstance(axes_attr, (list, numpy.ndarray)):
         axes_attr = [axes_attr]  # ensure it is array
 
-    for axis_name in utils.decode_byte_string(v_item.h5_object):
+    for axis_name in axes_attr:
         # check each value of array that is a validItemName
         k = item_name.validItemName_match_key(validator, axis_name)
         test_name = f"valid name @axes['{axis_name}']"

